builtin/aws/lambda/releaser.go

```diff
@@ -0,0 +1,323 @@
+package lambda
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/lambda"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint-plugin-sdk/docs"
+	"github.com/hashicorp/waypoint-plugin-sdk/framework/resource"
+	sdk "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+	"github.com/hashicorp/waypoint-plugin-sdk/terminal"
+	"github.com/hashicorp/waypoint/builtin/aws/utils"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type Releaser struct {
+	config ReleaserConfig
+}
+
+// Config implements Configurable
+func (r *Releaser) Config() (interface{}, error) {
+	return &r.config, nil
+}
+
+// ReleaseFunc implements component.ReleaseManager
+func (r *Releaser) ReleaseFunc() interface{} {
+	return r.Release
+}
+
+// StatusFunc implements component.Status
+func (r *Releaser) StatusFunc() interface{} {
+	return r.Status
+}
+
+// DestroyFunc implements component.Destroyer
+func (r *Releaser) DestroyFunc() interface{} {
+	return r.Destroy
+}
+
+func (r *Releaser) resourceManager(log hclog.Logger) *resource.Manager {
+	return resource.NewManager(
+		resource.WithLogger(log.Named("resource_manager")),
+		resource.WithValueProvider(r.getSession),
+
+		resource.WithResource(resource.NewResource(
+			resource.WithName("function_url"),
+			resource.WithState(&Resource_FunctionUrl{}),
+			resource.WithCreate(r.resourceFunctionUrlCreate),
+		)),
+	)
+}
+
+func (r *Releaser) getSession(
+	_ context.Context,
+	log hclog.Logger,
+) (*session.Session, error) {
+	return utils.GetSession(&utils.SessionConfig{
+		Logger: log,
+	})
+}
+
+var (
+	DefaultFunctionUrlAuthType = lambda.FunctionUrlAuthTypeNone
+)
+
+func (r *Releaser) resourceFunctionUrlCreate(
+	ctx context.Context,
+	log hclog.Logger,
+	sess *session.Session,
+	sg terminal.StepGroup,
+	ui terminal.UI,
+	dep *Deployment,
+	state *Resource_FunctionUrl,
+) error {
+	lambdasrv := lambda.New(sess)
+
+	functionUrlAuthType := DefaultFunctionUrlAuthType
+	if r.config.AuthType != "" {
+		functionUrlAuthType = strings.ToUpper(r.config.AuthType)
+	}
+
+	// TODO(thiskevinwang): source cors from HCL config
+	cors := lambda.Cors{}
+
+	addPermissionInput := lambda.AddPermissionInput{
+		Action:              aws.String("lambda:InvokeFunctionUrl"),
+		FunctionUrlAuthType: aws.String(functionUrlAuthType),
+		FunctionName:        aws.String(dep.FuncArn),
+		Principal:           aws.String("*"),
```

Comment:
Do you think we should make this configurable? We can have `Principal` default to `"*"` with the option to override in a config if necessary.

Reply:
Yea I think you're right — it should be configurable.

I think a `principal` config should only be applicable only when AuthType is `AWS_IAM` though (this looks like how the dashboard option behaves)

Reply:
I also realized I need to rework the permission creation logic.

Lambda unfortunately doesn't have an `edit-permission` API, only an [`add-permission`](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/lambda/add-permission.html) and [`remove-permission`](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/lambda/remove-permission.html).

...The AWS console performs a _remove-then-add_ when you **edit** a permission.

I think the simplest approach is to use a fixed `statementId` and **always**...
- remove-permission
  - no-op if not-found [^1]
- add-permission
... on every `waypoint up`

[^1]: This accounts for the very first `waypoint up` where a permission doesn't exist yet. It also accounts for a state where a permission was removed, and the following add-operation fails, leave you with no permission.

Alternatively, we could [`get-policy`](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/lambda/get-policy.html) first to find & determine if permissions need to be created or updated at all... but this looks like it gets unnecessarily complex very quick

Reply:
Ok, I took a stab at the above logic (the "Alternatively..." approach) - https://github.com/hashicorp/waypoint/pull/3187/commits/5f9f3fdd5cc6ad6b4082fe1f2eb6cde6740c8841

At least with my local testing, it works quite well with the various permutations of `auth_type` and `principal`:D

```hcl
# full public access
auth_type = "NONE" // can be omitted
// principal       // must be omitted, otherwise validation error 

# any authenticated AWS user, from any account
auth_type = "AWS_IAM"
principal = "*" // can be omitted

# any principal under <AWS_ACCOUNT_ID>... that is also permitted to perform `lambda:InvokeFunctionUrl`
auth_type = "AWS_IAM"
principal = <AWS_ACCOUNT_ID>

```

Reply:
This is a great question! I will bring this up in our next GitHub triage sync and let you know this week 👍🏻 Appreciate you taking the time to look into this and laying out the options. Makes it easier for us!

Reply:
Ok, we think the best course of action here is to make the default value for auth type to be `"NONE"`, and then let folks configure it and principal by request.

Reply:
@briancain if I'm understanding this correctly, then that means there are no further code changes required.. (?)

Reply:
I think you're right! Looks good, the config validator will do the check for us based on what you mentioned above for the right use-cases for auth, so looks good! Resolved. Thanks :)